### PR TITLE
🎉 (line+slope) add focus state / TAS-739

### PR DESCRIPTION
**Pull request by sophiamersmann**
Adds the option to highlight lines in line and slope charts

### Summary

- Adds a new config field, `focusedSeriesNames` (suggestions for a different name welcome)
- A line is identified by its series name which is either an entity name, a column display name or a combination of both
- The list of focused series names is persisted to the URL as `focus` query param
    - The entity name utility functions are used to parse and serialise focused series names, so that the same delimiter is used and entity names are mapped to their codes if possible
    - This breaks if a column name contains `~` (the delimiter) which theoretically is possible but I don't think we need to worry about that now
- Focused lines have bold labels, non-focused lines are grayed out
- Grapher makes an effort to prevent the chart to enter a 'bad state' where all lines are grayed out because the focused line doesn't exist
    - This includes removing all elements from the focus array when the facet strategy changes and dismissing focused entities when they're unselected

#### In the admin

- There is a new 'Data to highlight' section below the entity selection section
- If the chart is in a bad state because one of the focused series names is invalid, saving is disabled and shows an error message

### Follow up

- It's a bit ugly that `selectedEntityNames` and `focusedSeriesNames` are always serialised, even for an empty Grapher. I've fixes that in a [follow-up PR](https://github.com/owid/owid-grapher/pull/4294)
- The line legend method that drops labels if there are to many is a bit difficult to read. I'll open another PR with a refactor

### CODE REVIEW

**Comment by sophiamersmann** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4272?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4272** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4272?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4272?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#4266** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4266?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4207** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4207?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#4210](https://github.com/owid/owid-grapher/pull/4210) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4210?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#4243](https://github.com/owid/owid-grapher/pull/4243) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4243?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#4247** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4247?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4222** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4222?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4205** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4192** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4192?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4170** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4159** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4159?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by marcelgerber — APPROVED** [review]
Nice! I've tested this a bunch in some line & slope charts, and it all works well overall.

A few observations:
- I find it a bit weird that for slope charts, clicking the right legend entry changes focus state, but clicking the left one (or the line itself) does nothing.
- It would be nice if discrete bar charts could also support this.

> Grapher makes an effort to prevent the chart to enter a 'bad state' where all lines are grayed out because the focused line doesn't exist

There's one such case that is still possible, which you can see in http://staging-site-focus-state-viz/grapher/population-density-of-the-capital-city?country=~DEU&focus=~FRA: Here, `selectedEntityNames` and `focusedSeriesNames` are non-overlapping, making everything gray. Although it should be quite hard to get into such a state without manually changing the URL, if I see that correctly?

> The line legend method that drops labels if there are to many is a bit difficult to read. I'll open another PR with a refactor

Yeah I agree, it's become quite overwhelming by now.

**Comment by sophiamersmann** [comment]
Thank you for the quick review!

> - I find it a bit weird that for slope charts, clicking the right legend entry changes focus state, but clicking the left one (or the line itself) does nothing.

true! will add it for the left legend. I don't want to overload slope interactions too much though and since it's only a power user feature...

> - It would be nice if discrete bar charts could also support this.

yap!

> There's one such case that is still possible, which you can see in http://staging-site-focus-state-viz/grapher/population-density-of-the-capital-city?country=~DEU&focus=~FRA: Here, `selectedEntityNames` and `focusedSeriesNames` are non-overlapping, making everything gray. Although it should be quite hard to get into such a state without manually changing the URL, if I see that correctly?

The URL isn't validated at all. You can also focus a non-existing entity and everything is grayed out: http://staging-site-focus-state-viz/grapher/population-density-of-the-capital-city?focus=hello. That's how Grapher currently works for other query params as well, so I just went with that for now :)

File: packages/@ourworldindata/grapher/src/core/Grapher.tsx [hunk 1]
```diff
@@ -2783,6 +2818,11 @@ export class Grapher
         this.selectedFacetStrategy = facet
     }
 
+    @computed get isFaceted(): boolean {
+        const hasFacetStrategy = this.facetStrategy !== FacetStrategy.none
+        return !this.isOnMapTab && hasFacetStrategy
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply packages/@ourworldindata/grapher/src/core/Grapher.tsx:2823]: yeah, probably!

**Comment by sophiamersmann** [comment]
### Merge activity

* **Dec 17, 11:23 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/4272).

**Pull request merged**